#Домашнє завдання до теми "Comprehensions в Python"

Для наступних завдань у нас є наступний список:

In [1]:
numbers = [42, 51, 29, 2, 13, 44, 80, 66, 16]

1. Використовуючи цикл `for` cтворіть новий список `numbers_squared`, де кожен елемент списку `numbers` є домноженим на 2.

Нагадаю, що фактично коли ми з допомогою циклу отримуємо новий список, то ми спочатку створюємо пустий список, а потім додаємо в нього новий елемент на кожній ітерації.

Новий елемент буде в даному випадку результатом домножання чергового елемента вихідного списку `numbers` на 2.

In [7]:
numbers_squared = []
for number in numbers:
    numbers_squared.append(number * 2)
numbers_squared

[84, 102, 58, 4, 26, 88, 160, 132, 32]

Очікуваний результат: `[84, 102, 58, 4, 26, 88, 160, 132, 32]`

2. Використовуючи `list comprehension` cтворіть новий список `numbers_squared_lh`, де кожен елемент списку `numbers` є домноженим 2.
Так, операція, яку ми викнуємо над кожним елементом вихідного списку, та сама, що і в попередньому завданні, але інший спосіб отримання результату.

In [4]:
numbers_squared_lh = [number * 2 for number in numbers]
numbers_squared_lh

[84, 102, 58, 4, 26, 88, 160, 132, 32]

3. У нас вже є знайдене середнє значення за списком - воно визначене в змінній `numbers_mean`.

  Використовуючи цикл, створіть новий список `numbers_gt_mean`, який порівнює кожен елемент з цим середнім значенням за списком за наступним правилом: "якщо цей елемент спиcку `numbers` є більшим за `numbers_mean`, то запишемо 1, інакше 0". Виведіть новий список на екран.

In [8]:
numbers_mean = 38.11

In [9]:
numbers_gt_mean = []
for number in numbers:
    if number > numbers_mean:
        numbers_gt_mean.append(1)
    else:
        numbers_gt_mean.append(0)
numbers_gt_mean

[1, 1, 0, 0, 0, 1, 1, 1, 0]

Очікуваний результат: `[1, 1, 0, 0, 0, 1, 1, 1, 0]`

4. Отримайте список за тим самим правилом, що в завданні 3, тільки тепер з list comprehension. Результат запишіть у змінну `numbers_gt_mean_lc`

In [11]:
numbers_gt_mean_lc = [1 if number > numbers_mean else 0 for number in numbers]
numbers_gt_mean_lc

[1, 1, 0, 0, 0, 1, 1, 1, 0]

5. Заданий список `long`. З допомогою циклу створіть словник `n_repeats`, який містить інформацію "скільки разів повторюється кожен унікальний елемент списку `long`".
Наприклад:
```
long = [1, 1, 2, 2, 3]
# тоді `n_repeats` містить {1: 2, 2: 2, 3: 1}
```

Аби для кожного унікального елементу знайти, скільки разів він повторюється, виористайте метод списку `list.count(<елемент>)`.
Виведіть `n_repeats` на екран.

**Підказки**:
1. Для початку може бути зручно отримати набір унікальних елементів списку.
2. Порядок ключів в словнику не має значення! Тобто ключі в словнику не обовʼязково мають бути в тому ж порядку, як наведено в прикладі результату. Але кількості зустрічань для кожного елемента мають співпадати з наведеними.

In [12]:
long = [75, 39, 95, 1, 91, 43, 97, 24, 86, 75, 92, 91, 15, 15, 75, 95, 54, 29, 55, 98]

In [19]:
long_set = set(long)
n_repeats = {}
for elem in long_set:
    n_repeats[elem] = long.count(elem)
print(n_repeats)


{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}


Очікуваний результат:
```
{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}

```

6. Виконайте ту саму операцію, що і в попередньому заваднні, але використовуючи `dict comprehension`. Результат запишіть у змінну `n_repeats_dh` та результат виведіть на екран.

In [23]:
n_repeats_dh = {element: long.count(element) for element in long_set}
print(n_repeats_dh)

{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}


7. Знайдіть три найменші **унікальні** елементи в списку `long` та виведіть їх на екран.

In [40]:
long = [75, 39, 95, 1, 91, 43, 97, 24, 86, 75, 92, 91, 15, 15, 75, 95, 54, 29, 55, 98, 1, 1]

In [47]:
sort_long = sorted(long)
unique_first_el = [sort_long[0]]
for i, el in enumerate(sort_long):
    if el > sort_long[i-1] and len(unique_first_el) <=2:
        unique_first_el.append(el)
unique_first_el

# купа різних ідей і реалізація без жодних підказок! Думаю тут багатьма способами можна досягнути результату (бо ідей в голові багато, а знань ще не дуже)

[1, 15, 24]

Очікувана відповідь: [1, 15, 24]

In [49]:
# Другий спосіб
unique_values = []
for value in long:
    if value not in unique_values:
        unique_values.append(value)
unique_values.sort()
three_smallest = unique_values[:3]
print(three_smallest)

[1, 15, 24]


In [58]:
# Третій спосіб
unique_values2 = [val for val in long if val not in unique_values2]
unique_values2.sort()
# three_smallest2 = unique_values2[:3]
print(unique_values2)
# Так чомусь додає в список однакові елементи.

[1, 1, 1, 15, 15, 24, 29, 39, 43, 54, 55, 75, 75, 75, 86, 91, 91, 92, 95, 95, 97, 98]


In [67]:
# Четвертий спосіб
long_set = set(long)
unique_list = list(long_set)
three_smallest_el = sorted(unique_list)
three_smallest_el[:3]

[1, 15, 24]